tests/integration/test_upgrades.py

```diff
@@ -0,0 +1,104 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Ranger charm upgrades integration tests."""
+
+import asyncio
+import logging
+
+import pytest
+import pytest_asyncio
+import requests
+import yaml
+from helpers import APP_NAME, METADATA, POSTGRES_NAME, SECURE_PWD, get_unit_url
+from pytest_operator.plugin import OpsTest
+
+logger = logging.getLogger(__name__)
+
+
+@pytest.mark.skip_if_deployed
+@pytest_asyncio.fixture(name="deploy", scope="module")
+async def deploy(ops_test: OpsTest):
+    """Deploy the app."""
+    asyncio.gather(
+        ops_test.model.deploy(POSTGRES_NAME, channel="14", trust=True),
+    )
+
+    async with ops_test.fast_forward():
```

Comment:
Would you save time if you made the deploy calls first, then waited for idle states instead of interleaving them?

Reply:
Yes, that makes more sense timewise. Updated